File: 0_CONTEXT/Computer_Languages/PHP/ActiveInferenceAgent.php
<?php

/**
 * Active Inference Agent Implementation in PHP
 *
 * This class implements the core active inference algorithm using PHP's
 * object-oriented capabilities and array operations.
 */

class ActiveInferenceAgent {
    private int $nStates;
    private int $nObservations;
    private int $nActions;
    private array $beliefs;
    private array $priorBeliefs;
    private array $aMatrix;  // P(o|s) - observation likelihood
    private array $bMatrix;  // P(s'|s,a) - state transitions
    private array $cVector;  // P(o) - observation preferences
    private float $precision;
    private float $learningRate;

    // History for analysis
    private array $beliefHistory = [];
    private array $actionHistory = [];
    private array $observationHistory = [];
    private array $freeEnergyHistory = [];

    /**
     * Constructor
     */
    public function __construct(int $nStates, int $nObservations, int $nActions,
                               float $precision = 1.0, float $learningRate = 0.1) {
        $this->nStates = $nStates;
        $this->nObservations = $nObservations;
        $this->nActions = $nActions;
        $this->precision = $precision;
        $this->learningRate = $learningRate;

        $this->initializeGenerativeModel();
        $this->beliefs = $this->priorBeliefs;
        $this->beliefHistory[] = $this->beliefs;
    }

    /**
     * Initialize the generative model matrices
     */
    private function initializeGenerativeModel(): void {
        // Initialize A matrix (observation likelihood)
        $this->aMatrix = [];
        for ($o = 0; $o < $this->nObservations; $o++) {
            $this->aMatrix[$o] = [];
            for ($s = 0; $s < $this->nStates; $s++) {
                // Diagonal structure with some noise
                if (($s % $this->nObservations) === $o) {
                    $this->aMatrix[$o][$s] = 0.8;
                } else {
                    $this->aMatrix[$o][$s] = 0.2 / ($this->nObservations - 1);
                }
            }
        }

        // Initialize B matrix (state transitions)
        $this->bMatrix = [];
        for ($a = 0; $a < $this->nActions; $a++) {
            $this->bMatrix[$a] = [];
            for ($s = 0; $s < $this->nStates; $s++) {
                $this->bMatrix[$a][$s] = [];
                for ($sNext = 0; $sNext < $this->nStates; $sNext++) {
                    // Action-specific transition patterns
                    if ($sNext === (($s + $a + 1) % $this->nStates)) {
                        $this->bMatrix[$a][$s][$sNext] = 0.7;
                    } elseif ($sNext === $s) {
                        $this->bMatrix[$a][$s][$sNext] = 0.2;
                    } else {
                        $this->bMatrix[$a][$s][$sNext] = 0.1 / ($this->nStates - 2);
                    }
                }
            }
        }

        // Initialize C vector (observation preferences)
        $this->cVector = [];
        for ($o = 0; $o < $this->nObservations; $o++) {
            $this->cVector[$o] = ($o === 0) ? 0.0 : 0.5;
        }

        // Initialize D vector (prior beliefs)
        $this->priorBeliefs = [];
        for ($s = 0; $s < $this->nStates; $s++) {
            $this->priorBeliefs[$s] = 1.0 / $this->nStates;
        }
    }

    /**
     * Update beliefs based on observation using Bayesian inference
     */
    public function updateBeliefs(int $observation): void {
        // Get likelihood for this observation
        $likelihood = $this->aMatrix[$observation];

        // Compute posterior: P(s|o) ∝ P(o|s) * P(s)
        $posterior = [];
        for ($s = 0; $s < $this->nStates; $s++) {
            $posterior[$s] = $likelihood[$s] * $this->beliefs[$s];
        }

        // Normalize
        $sum = array_sum($posterior);
        if ($sum > 0) {
            for ($s = 0; $s < $this->nStates; $s++) {
                $this->beliefs[$s] = $posterior[$s] / $sum;
            }
        }

        // Store in history
        $this->beliefHistory[] = $this->beliefs;
        $this->observationHistory[] = $observation;
    }

    /**
     * Select action by minimizing expected free energy
     */
    public function selectAction(): int {
        $minEfe = PHP_FLOAT_MAX;
        $bestAction = 0;

        for ($action = 0; $action < $this->nActions; $action++) {
            $predictedBeliefs = $this->predictBeliefs($action);
            $efe = $this->calculateExpectedFreeEnergy($predictedBeliefs);

            if ($efe < $minEfe) {
                $minEfe = $efe;
                $bestAction = $action;
            }
        }

        $this->actionHistory[] = $bestAction;
        return $bestAction;
    }

    /**
     * Predict beliefs after taking an action
     */
    private function predictBeliefs(int $action): array {
        $predictedBeliefs = array_fill(0, $this->nStates, 0.0);

        for ($sNext = 0; $sNext < $this->nStates; $sNext++) {
            for ($s = 0; $s < $this->nStates; $s++) {
                $predictedBeliefs[$sNext] += $this->bMatrix[$action][$s][$sNext] * $this->beliefs[$s];
            }
        }

        return $predictedBeliefs;
    }

    /**
     * Calculate expected free energy for predicted beliefs
     */
    private function calculateExpectedFreeEnergy(array $predictedBeliefs): float {
        $efe = 0.0;

        for ($s = 0; $s < $this->nStates; $s++) {
            if ($predictedBeliefs[$s] > 0) {
                // KL divergence between predicted beliefs and prior
                $efe += $predictedBeliefs[$s] * log($predictedBeliefs[$s] / $this->priorBeliefs[$s]);
            }
        }

        return $efe * $this->precision;
    }

    /**
     * Calculate current variational free energy
     */
    public function calculateFreeEnergy(): float {
        $fe = 0.0;

        for ($s = 0; $s < $this->nStates; $s++) {
            if ($this->beliefs[$s] > 0) {
                $fe += $this->beliefs[$s] * log($this->beliefs[$s] / $this->priorBeliefs[$s]);
            }
        }

        $this->freeEnergyHistory[] = $fe;
        return $fe;
    }

    /**
     * Execute one perception-action cycle
     */
    public function step(int $observation): int {
        $this->updateBeliefs($observation);
        $action = $this->selectAction();
        $this->calculateFreeEnergy();
        return $action;
    }

    /**
     * Learn from experience
     */
    public function learn(int $observation, int $action, int $nextObservation): void {
        // Simple learning: reinforce the most likely state
        $maxState = array_keys($this->beliefs, max($this->beliefs))[0];

        // Update A matrix
        $this->aMatrix[$observation][$maxState] += $this->learningRate;

        // Renormalize row
        $rowSum = array_sum($this->aMatrix[$observation]);
        for ($s = 0; $s < $this->nStates; $s++) {
            $this->aMatrix[$observation][$s] /= $rowSum;
        }
    }

    /**
     * Calculate belief entropy
     */
    public function calculateBeliefEntropy(): float {
        $entropy = 0.0;
        for ($s = 0; $s < $this->nStates; $s++) {
            if ($this->beliefs[$s] > 0) {
                $entropy -= $this->beliefs[$s] * log($this->beliefs[$s]);
            }
        }
        return $entropy;
    }

    /**
     * Get current beliefs
     */
    public function getBeliefs(): array {
        return $this->beliefs;
    }

    /**
     * Get belief history
     */
    public function getBeliefHistory(): array {
        return $this->beliefHistory;
    }

    /**
     * Print agent statistics
     */
    public function printStatistics(): void {
        echo "Active Inference Agent Statistics:\n";
        echo "=====================================\n";
        echo "States: {$this->nStates}, Observations: {$this->nObservations}, Actions: {$this->nActions}\n";
        echo "Precision: {$this->precision}, Learning Rate: {$this->learningRate}\n";

        if (!empty($this->actionHistory)) {
            echo "Total steps: " . count($this->actionHistory) . "\n";
            echo "Final beliefs: [";
            for ($s = 0; $s < $this->nStates; $s++) {
                echo number_format($this->beliefs[$s], 3);
                if ($s < $this->nStates - 1) {
                    echo ", ";
                }
            }
            echo "]\n";

            if (!empty($this->freeEnergyHistory)) {
                $finalFe = end($this->freeEnergyHistory);
                echo "Final free energy: " . number_format($finalFe, 4) . "\n";
            }

            $entropy = $this->calculateBeliefEntropy();
            echo "Belief entropy: " . number_format($entropy, 4) . "\n";
        } else {
            echo "No steps taken yet.\n";
        }
    }

    /**
     * Get comprehensive statistics
     */
    public function getStatistics(): array {
        $stats = [
            'n_states' => $this->nStates,
            'n_observations' => $this->nObservations,
            'n_actions' => $this->nActions,
            'precision' => $this->precision,
            'learning_rate' => $this->learningRate,
            'current_beliefs' => $this->beliefs,
            'total_steps' => count($this->actionHistory),
            'belief_history_length' => count($this->beliefHistory)
        ];

        if (!empty($this->freeEnergyHistory)) {
            $stats['final_free_energy'] = end($this->freeEnergyHistory);
            $stats['free_energy_history'] = $this->freeEnergyHistory;
        }

        if (!empty($this->actionHistory)) {
            $stats['action_history'] = $this->actionHistory;
            $stats['observation_history'] = $this->observationHistory;
        }

        $stats['belief_entropy'] = $this->calculateBeliefEntropy();

        return $stats;
    }
}

/**
 * Demonstration class
 */
class ActiveInferenceDemo {
    public static function run(): void {
        echo "🧠 Active Inference PHP Implementation\n";
        echo "=====================================\n\n";

        // Create agent
        $nStates = 3;
        $nObservations = 3;
        $nActions = 2;

        echo "Creating agent with {$nStates} states, {$nObservations} observations, {$nActions} actions...\n\n";

        $agent = new ActiveInferenceAgent($nStates, $nObservations, $nActions, 1.0, 0.1);

        // Print initial beliefs
        echo "Initial beliefs: [";
        $beliefs = $agent->getBeliefs();
        for ($s = 0; $s < count($beliefs); $s++) {
            echo number_format($beliefs[$s], 3);
            if ($s < count($beliefs) - 1) {
                echo ", ";
            }
        }
        echo "]\n\n";

        // Run simulation
        $nSteps = 10;
        echo "Running simulation for {$nSteps} steps...\n\n";

        for ($step = 1; $step <= $nSteps; $step++) {
            // Generate random observation
            $observation = random_int(1, $nObservations);

            echo "Step {$step}:\n";
            echo "  Observation: {$observation}\n";

            // Execute perception-action cycle
            $action = $agent->step($observation);

            echo "  Selected action: {$action}\n";

            // Print current beliefs
            $currentBeliefs = $agent->getBeliefs();
            echo "  Updated beliefs: [";
            for ($s = 0; $s < count($currentBeliefs); $s++) {
                echo number_format($currentBeliefs[$s], 3);
                if ($s < count($currentBeliefs) - 1) {
                    echo ", ";
                }
            }
            echo "]\n";

            // Print free energy
            $stats = $agent->getStatistics();
            if (isset($stats['final_free_energy'])) {
                echo "  Free energy: " . number_format($stats['final_free_energy'], 4) . "\n";
            }

            echo "\n";
        }

        // Print final statistics
        echo "\n";
        $agent->printStatistics();

        echo "\n✅ PHP Active Inference simulation completed!\n";
    }

    /**
     * Multi-agent simulation demo
     */
    public static function runMultiAgentDemo(): void {
        echo "\n🐜 Multi-Agent PHP Demo\n";
        echo "======================\n\n";

        $nAgents = 3;
        $agents = [];

        // Create multiple agents
        for ($i = 0; $i < $nAgents; $i++) {
            $agents[] = new ActiveInferenceAgent(3, 3, 2, 1.0, 0.1);
        }

        // Run multi-agent simulation
        for ($step = 1; $step <= 5; $step++) {
            echo "Multi-agent Step {$step}:\n";

            for ($i = 0; $i < $nAgents; $i++) {
                $observation = random_int(1, 3);
                $action = $agents[$i]->step($observation);

                $beliefs = $agents[$i]->getBeliefs();
                $maxBelief = max($beliefs);

                echo "  Agent " . ($i + 1) . ": Obs={$observation}, Action={$action}, ";
                echo "Best Belief=" . number_format($maxBelief, 3) . "\n";
            }
            echo "\n";
        }

        echo "✅ Multi-agent simulation completed!\n";
    }
}

// Run the demo if this file is executed directly
if (basename(__FILE__) === basename($_SERVER['PHP_SELF'] ?? __FILE__)) {
    ActiveInferenceDemo::run();
    ActiveInferenceDemo::runMultiAgentDemo();
}
?>

File: 0_CONTEXT/Computer_Languages/Ruby/active_inference_agent.rb
#!/usr/bin/env ruby

# Active Inference Implementation in Ruby
# =====================================
#
# This implementation demonstrates the Active Inference framework using Ruby's
# elegant object-oriented design and array operations.

require 'matrix'

class ActiveInferenceAgent
  attr_reader :n_states, :n_observations, :n_actions, :precision, :learning_rate
  attr_reader :beliefs, :belief_history, :action_history, :observation_history, :free_energy_history

  def initialize(n_states: 4, n_observations: 3, n_actions: 2, precision: 1.0, learning_rate: 0.1)
    @n_states = n_states
    @n_observations = n_observations
    @n_actions = n_actions
    @precision = precision
    @learning_rate = learning_rate

    # Initialize generative model
    initialize_generative_model

    # Initialize beliefs to prior
    @beliefs = @d_vector.clone
    @belief_history = [@beliefs.clone]
    @action_history = []
    @observation_history = []
    @free_energy_history = []
  end

  def initialize_generative_model
    # Initialize A matrix (observation likelihood) - P(o|s)
    @a_matrix = Array.new(@n_observations) do |o|
      Array.new(@n_states) do |s|
        # Diagonal structure with some noise
        if (s % @n_observations) == o
          0.8  # High probability for matching observation
        else
          0.2 / (@n_observations - 1)  # Low probability for others
        end
      end
    end

    # Initialize B matrix (state transitions) - P(s'|s,a)
    @b_matrix = Array.new(@n_actions) do |a|
      Array.new(@n_states) do |s|
        Array.new(@n_states) do |s_next|
          # Action-specific transition patterns
          if s_next == ((s + a + 1) % @n_states)
            0.7  # Likely to move forward by action amount
          elsif s_next == s
            0.2  # Some probability of staying
          else
            0.1 / (@n_states - 2)  # Low probability for others
          end
        end
      end
    end

    # Initialize C vector (observation preferences) - P(o)
    @c_vector = Array.new(@n_observations) do |o|
      o == 0 ? 0.0 : 0.5  # Lower energy for preferred observation
    end

    # Initialize D vector (prior beliefs) - P(s)
    @d_vector = Array.new(@n_states) { 1.0 / @n_states }  # Uniform prior
  end

  # Update beliefs using Bayesian inference
  def update_beliefs(observation)
    # Get likelihood for this observation
    likelihood = @a_matrix[observation]

    # Compute posterior: P(s|o) ∝ P(o|s) * P(s)
    posterior = likelihood.zip(@beliefs).map { |l, b| l * b }

    # Normalize
    total = posterior.sum
    @beliefs = total > 0 ? posterior.map { |p| p / total } : posterior

    # Store in history
    @belief_history << @beliefs.clone
    @observation_history << observation
  end

  # Select action by minimizing expected free energy
  def select_action
    min_efe = Float::INFINITY
    best_action = 0

    @n_actions.times do |action|
      predicted_beliefs = predict_beliefs(action)
      efe = calculate_expected_free_energy(predicted_beliefs)

      if efe < min_efe
        min_efe = efe
        best_action = action
      end
    end

    @action_history << best_action
    best_action
  end

  # Predict beliefs after taking an action
  def predict_beliefs(action)
    predicted_beliefs = Array.new(@n_states, 0.0)

    @n_states.times do |s_next|
      @n_states.times do |s|
        predicted_beliefs[s_next] += @b_matrix[action][s][s_next] * @beliefs[s]
      end
    end

    predicted_beliefs
  end

  # Calculate expected free energy for predicted beliefs
  def calculate_expected_free_energy(predicted_beliefs)
    efe = 0.0

    @n_states.times do |s|
      if predicted_beliefs[s] > 0
        # KL divergence between predicted beliefs and prior
        efe += predicted_beliefs[s] * Math.log(predicted_beliefs[s] / @d_vector[s])
      end
    end

    efe * @precision
  end

  # Calculate current variational free energy
  def calculate_free_energy
    fe = 0.0

    @n_states.times do |s|
      if @beliefs[s] > 0
        fe += @beliefs[s] * Math.log(@beliefs[s] / @d_vector[s])
      end
    end

    @free_energy_history << fe
    fe
  end

  # Execute one perception-action cycle
  def step(observation)
    update_beliefs(observation)
    action = select_action
    calculate_free_energy
    action
  end

  # Learn from experience
  def learn(observation, action, next_observation)
    # Simple learning: reinforce the most likely state
    max_state = @beliefs.each_with_index.max[1]

    # Update A matrix
    @a_matrix[observation][max_state] += @learning_rate

    # Renormalize row
    row_sum = @a_matrix[observation].sum
    @a_matrix[observation].map! { |prob| prob / row_sum }
  end

  # Calculate belief entropy
  def calculate_belief_entropy
    entropy = 0.0
    @n_states.times do |s|
      if @beliefs[s] > 0
        entropy -= @beliefs[s] * Math.log(@beliefs[s])
      end
    end
    entropy
  end

  # Print agent statistics
  def print_statistics
    puts "Active Inference Agent Statistics:"
    puts "====================================="
    puts "States: #{@n_states}, Observations: #{@n_observations}, Actions: #{@n_actions}"
    puts "Precision: #{format('%.2f', @precision)}, Learning Rate: #{format('%.2f', @learning_rate)}"

    if !@action_history.empty?
      puts "Total steps: #{@action_history.length}"
      print "Final beliefs: ["
      @n_states.times do |s|
        print format('%.3f', @beliefs[s])
        print ', ' if s < @n_states - 1
      end
      puts ']'

      if !@free_energy_history.empty?
        puts "Final free energy: #{format('%.4f', @free_energy_history.last)}"
      end

      entropy = calculate_belief_entropy
      puts "Belief entropy: #{format('%.4f', entropy)}"
    else
      puts "No steps taken yet."
    end
  end

  # Get comprehensive statistics
  def get_statistics
    stats = {
      n_states: @n_states,
      n_observations: @n_observations,
      n_actions: @n_actions,
      precision: @precision,
      learning_rate: @learning_rate,
      current_beliefs: @beliefs.clone,
      total_steps: @action_history.length,
      belief_history_length: @belief_history.length
    }

    if !@free_energy_history.empty?
      stats[:final_free_energy] = @free_energy_history.last
      stats[:free_energy_history] = @free_energy_history.clone
    end

    if !@action_history.empty?
      stats[:action_history] = @action_history.clone
      stats[:observation_history] = @observation_history.clone
    end

    stats[:belief_entropy] = calculate_belief_entropy
    stats
  end

  # Get current beliefs
  def get_beliefs
    @beliefs.clone
  end
end

# Demonstration class
class ActiveInferenceDemo
  def self.run
    puts "🧠 Active Inference Ruby Implementation"
    puts "======================================"
    puts

    # Create agent
    n_states = 3
    n_observations = 3
    n_actions = 2

    puts "Creating agent with #{n_states} states, #{n_observations} observations, #{n_actions} actions..."
    puts

    agent = ActiveInferenceAgent.new(
      n_states: n_states,
      n_observations: n_observations,
      n_actions: n_actions,
      precision: 1.0,
      learning_rate: 0.1
    )

    # Print initial beliefs
    puts "Initial beliefs: [#{agent.get_beliefs.map { |b| format('%.3f', b) }.join(', ')}]"
    puts

    # Run simulation
    n_steps = 10
    puts "Running simulation for #{n_steps} steps..."
    puts

    n_steps.times do |step|
      # Generate random observation
      observation = rand(n_observations)

      puts "Step #{step + 1}:"
      puts "  Observation: #{observation}"

      # Execute perception-action cycle
      action = agent.step(observation)

      puts "  Selected action: #{action}"

      # Print current beliefs
      current_beliefs = agent.get_beliefs
      puts "  Updated beliefs: [#{current_beliefs.map { |b| format('%.3f', b) }.join(', ')}]"

      # Print free energy
      stats = agent.get_statistics
      if stats.key?(:final_free_energy)
        puts "  Free energy: #{format('%.4f', stats[:final_free_energy])}"
      end

      puts
    end

    # Print final statistics
    puts
    agent.print_statistics

    puts
    puts "✅ Ruby Active Inference simulation completed!"
  end

  def self.run_multi_agent_demo
    puts
    puts "🐜 Multi-Agent Ruby Demo"
    puts "========================"
    puts

    n_agents = 3
    agents = []

    # Create multiple agents
    n_agents.times do |i|
      agents << ActiveInferenceAgent.new(
        n_states: 3,
        n_observations: 3,
        n_actions: 2,
        precision: 1.0,
        learning_rate: 0.1
      )
    end

    # Run multi-agent simulation
    5.times do |step|
      puts "Multi-agent Step #{step + 1}:"

      agents.each_with_index do |agent, i|
        observation = rand(3)
        action = agent.step(observation)

        beliefs = agent.get_beliefs
        max_belief = beliefs.max

        puts "  Agent #{i + 1}: Obs=#{observation}, Action=#{action}, " +
             "Best Belief=#{format('%.3f', max_belief)}"
      end
      puts
    end

    puts "✅ Multi-agent simulation completed!"
  end
end

# Learning demonstration
class LearningDemo
  def self.run
    puts
    puts "🎓 Active Inference Learning Demo"
    puts "================================"
    puts

    agent = ActiveInferenceAgent.new(
      n_states: 4,
      n_observations: 3,
      n_actions: 2,
      precision: 1.0,
      learning_rate: 0.2
    )

    puts "Initial A-matrix (observation model):"
    agent.instance_variable_get(:@a_matrix).each_with_index do |row, o|
      puts "  O#{o}: [#{row.map { |p| format('%.2f', p) }.join(', ')}]"
    end
    puts

    # Learning loop
    20.times do |trial|
      # Generate observation
      observation = rand(3)

      # Step through environment
      action = agent.step(observation)

      # Generate next observation based on action
      next_observation = (observation + action + rand(2)) % 3

      # Learn from experience
      agent.learn(observation, action, next_observation)
    end

    puts "After learning:"
    agent.instance_variable_get(:@a_matrix).each_with_index do |row, o|
      puts "  O#{o}: [#{row.map { |p| format('%.2f', p) }.join(', ')}]"
    end

    puts
    agent.print_statistics

    puts
    puts "✅ Learning demonstration completed!"
  end
end

# Run demos if this file is executed directly
if __FILE__ == $0
  ActiveInferenceDemo.run
  ActiveInferenceDemo.run_multi_agent_demo
  LearningDemo.run
end
